### PR TITLE
Allow Kafka producer headers to be dict or list

### PR DESCRIPTION
**Pull request by mrajashree**
# Description

Fixes https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1364

The package confluent-kafka has a class `Producer`, with a method `produce`, which accepts headers as one of the parameters, and this `headers` parameter can be a list or a dict. (Ref: https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.Producer.produce)

However, passing a dictionary of headers to Producer when using opentelemetry-instrumentation-confluent-kafka doesn't work, because the KafkaContextSetter's set method tries to append the headers assuming the carrier is a list.

This PR checks the carrier type, and appends the `(key, value)` tuple in case of a list, or adds the `key:value` to the dict in case the carrier is a dict.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- I was passing a dictionary of headers to the producer, earlier it used to crash since the code was calling `append` on a dict. With the changes in this PR it appends the header in case of a list, or adds the header to the existing dict as expected.
- Added unit tests which test context_setter.set, and assert that the new headers get added to the carrier when it is a list and also when it is a dict.

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [ ] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by shalevr** [comment]
Thank you for your contribution.
Don't you need to change the KafkaContextGetter too?
The old implementation expect carrier to be a list

File: instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py:69]: I think you meant to check the context_setter carrier

File: instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py:73]: Same here

File: instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py:69]: After the context_setter.set the original carrier_dict changed?
or you meant to check the context_setter carrier?

File: instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py:73]: Same here

**Comment by shalevr** [comment]
Please add a changelog entry

File: instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py
Comment (mrajashree) [reply instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py:69]: yes that's what the assert is checking that the original `carrier_dict` has the `key2:val2` pair added on line 68

File: instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py
Comment (mrajashree) [reply instrumentation/opentelemetry-instrumentation-confluent-kafka/tests/test_instrumentation.py:73]: and yes here the test is ensuring that the new tuple being added on line 72 exists in the carrier_list (if that's your question)

**Comment by mrajashree** [comment]
@shalevr thanks for reviewing!

>Don't you need to change the KafkaContextGetter too?
The old implementation expect carrier to be a list

Since the KafkaContextGetter is being used on `record.headers()` [here](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/df32e8ca7f91282846020f27e9a8d351510a7b25/instrumentation/opentelemetry-instrumentation-confluent-kafka/src/opentelemetry/instrumentation/confluent_kafka/__init__.py#L336), and the [headers method](https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.Message.headers) returns a list of `(key, value)` tuples, the `get` method is working on the list as expected, so I don't think we need to change it

**Comment by mrajashree** [comment]
@srikanthccv I saw that you had approved the PR which then ran the workflows, but I had to make a change based on the lint output. I made that change and force pushed. Can you approve running the workflows on this again?

**Comment by mrajashree** [comment]
@srikanthccv @shalevr thanks for rebasing with the latest main commits. Do you think we can now merge this PR?

**Comment by srikanthccv** [comment]
It will be merged after it gets approvals, especially from the codeowner for the fixes.

File: CHANGELOG.md
Comment (srikanthccv) [inline CHANGELOG.md:53]: This should be moved to Unreleased fixed section now.

Comment (mrajashree) [reply CHANGELOG.md:53]: @srikanthccv I think that's where it is right now, if I view the markdown, it shows up in the Unreleased Fixed section: https://github.com/open-telemetry/opentelemetry-python-contrib/blob/739e1f0564b12e12788688d8de48adea3b3d7b2f/CHANGELOG.md#fixed

**Comment by mrajashree** [comment]
> @shalevr thanks for reviewing!
> 
> > Don't you need to change the KafkaContextGetter too?
> > The old implementation expect carrier to be a list
> 
> Since the KafkaContextGetter is being used on `record.headers()` [here](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/df32e8ca7f91282846020f27e9a8d351510a7b25/instrumentation/opentelemetry-instrumentation-confluent-kafka/src/opentelemetry/instrumentation/confluent_kafka/__init__.py#L336), and the [headers method](https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.Message.headers) returns a list of `(key, value)` tuples, the `get` method is working on the list as expected, so I don't think we need to change it

regarding this: The issue was only with the KafkaContextSetter.set, since it was being called on [headers which is passed](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/3bcc043e9dfc1b13421ff250fd941a40b306144d/instrumentation/opentelemetry-instrumentation-confluent-kafka/src/opentelemetry/instrumentation/confluent_kafka/__init__.py#L321) by the user and could be a dict or list. And KafkaContextGetter.get is only [being called on the output](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/3bcc043e9dfc1b13421ff250fd941a40b306144d/instrumentation/opentelemetry-instrumentation-confluent-kafka/src/opentelemetry/instrumentation/confluent_kafka/__init__.py#L340) of [record.headers() which always returns a list of tuples](https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.Message.headers), hence doesn’t need to check for a dict. But for consistency and to allow future use of this method on a dict and list, I’ve updated the get and keys method on KafkaContextGetter as well to handle both, a dict and list

**Comment by mrajashree** [comment]
> It will be merged after it gets approvals, especially from the codeowner for the fixes.

@srikanthccv okay thanks, who are the codeowners and can they be added to this PR’s reviewers list too? 

@shalevr since you first reviewed I've added a new commit, can you re-review to see if my changes/comments address your previous review comments? thanks

**Comment by srikanthccv** [comment]
No sure why actions didn't assign this author https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/.github/component_owners.yml#L27. 


cc @oxeye-dorkolog, @dorkolog

File: CHANGELOG.md
Comment (srikanthccv) [reply CHANGELOG.md:58]: There was a release on Friday, so the prev section is tagged and there is a new Unreleased section. Look at the diff.

Comment (mrajashree) [reply CHANGELOG.md]: oh I hadn't rebased with main yet, that's why it was pointing to the older commit...updated it now and moved to the new unreleased section

**Review by srikanthccv — APPROVED** [review]

**Review by shalevr — APPROVED** [review]

**Comment by srikanthccv** [comment]
@oxeye-dorkolog, @dorkolog another ping

**Pull request merged**
